smarts/core/traffic_history.py

```diff
@@ -0,0 +1,124 @@
+# Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+
+from cached_property import cached_property
+from contextlib import nullcontext, closing
+from functools import lru_cache
+import random
+import sqlite3
+
+
+class TrafficHistory:
+    def __init__(self, db):
+        self._db = db
+        self._db_cnxn = None
+
+    def connect_if_not(self):
+        if not self._db_cnxn:
+            self._db_cnxn = sqlite3.connect(self._db)
+
```

Comment:
I find calling connect on the traffic history to be somewhat odd. Perhaps just something closer to `initialize` or `setup`?

Reply:
Yeah, I considered `setup()`, but I shied away from that since calling it is optional.  This is an optimization to open a persistent connection if a client is going to be doing a large number multiple queries (which avoids the overhead of parsing the sqllite file for each one).  Other clients can call queries without using this and they'll also work.

Maybe something like `connect_for_repeated_queries()` paired with `disconnect()`?

Reply:
I see. Then in that case I understand why this would be needed.